relayer/relays/beefy/scanner.go

```diff
@@ -15,61 +15,113 @@ import (
 	results := make(chan ScanBlocksResult)
-	go scanBlocks(ctx, api, startBlock, results)
+	go scanBlocks(ctx, meta, api, startBlock, results)
 	return results, nil
 }
 
-func scanBlocks(ctx context.Context, api *gsrpc.SubstrateAPI, startBlock uint64, out chan<- ScanBlocksResult) {
+func scanBlocks(ctx context.Context, meta *types.Metadata, api *gsrpc.SubstrateAPI, startBlock uint64, out chan<- ScanBlocksResult) {
 	defer close(out)
 
-	sendError := func(err error) {
+	emitError := func(err error) {
 		select {
 		case <-ctx.Done():
 			return
 		case out <- ScanBlocksResult{Error: err}:
 		}
 	}
 
-	current := startBlock
-	for {
+	fetchFinalizedBeefyHeader := func() (*types.Header, error) {
 		finalizedHash, err := api.RPC.Beefy.GetFinalizedHead()
 		if err != nil {
-			sendError(fmt.Errorf("fetch finalized head: %w", err))
-			return
+			return nil, fmt.Errorf("fetch finalized head: %w", err)
 		}
 
 		finalizedHeader, err := api.RPC.Chain.GetHeader(finalizedHash)
 		if err != nil {
-			sendError(fmt.Errorf("fetch header for finalised head %v: %w", finalizedHash.Hex(), err))
-			return
+			return nil, fmt.Errorf("fetch header for finalised head %v: %w", finalizedHash.Hex(), err)
 		}
 
+		return finalizedHeader, nil
+	}
+
+	sessionCurrentIndexKey, err := types.CreateStorageKey(meta, "Session", "CurrentIndex", nil, nil)
+	if err != nil {
+		emitError(fmt.Errorf("create storage key: %w", err))
+		return
+	}
+
+	blockHash, err := api.RPC.Chain.GetBlockHash(max(startBlock-1, 0))
+	if err != nil {
+		emitError(fmt.Errorf("fetch block hash: %w", err))
+		return
+	}
+
+	// Get session index of block before start block
+	var currentSessionIndex uint32
+	_, err = api.RPC.State.GetStorage(sessionCurrentIndexKey, &currentSessionIndex, blockHash)
+	if err != nil {
+		emitError(fmt.Errorf("fetch session index: %w", err))
+		return
+	}
+
+	finalizedHeader, err := fetchFinalizedBeefyHeader()
+	if err != nil {
+		emitError(err)
+		return
+	}
+	current := startBlock
+	for {
+		log.Info("foo")
+
 		finalizedBlockNumber := uint64(finalizedHeader.Number)
 		if current > finalizedBlockNumber {
 			select {
 			case <-ctx.Done():
 				return
 			case <-time.After(3 * time.Second):
 			}
+			finalizedHeader, err = fetchFinalizedBeefyHeader()
+			if err != nil {
+				emitError(err)
+				return
+			}
 			continue
 		}
 
+		if current > uint64(finalizedHeader.Number) {
+			return
+		}
+
 		blockHash, err := api.RPC.Chain.GetBlockHash(current)
 		if err != nil {
-			sendError(fmt.Errorf("fetch block hash: %w", err))
+			emitError(fmt.Errorf("fetch block hash: %w", err))
+			return
+		}
+
+		var sessionIndex uint32
+		_, err = api.RPC.State.GetStorage(sessionCurrentIndexKey, &sessionIndex, blockHash)
+		if err != nil {
+			emitError(fmt.Errorf("fetch session index: %w", err))
 			return
 		}
 
+		if sessionIndex > currentSessionIndex {
+			log.Info("BOO")
+			currentSessionIndex = sessionIndex
+		} else {
+			current++
+			continue
+		}
```

Comment:
I don't think we need to get Session index to check whether it's handover commitment or not.

The request itself is enough to tell as `nextValidatorSetID(in proof) == validatorSetID(in commitment) + 1`

Reply:
Isn't that true for all commitments though?

Reply:
Sorry, it's `validatorSetID(in LC) < validatorSetID(in commitment)` then I'd assume it can be treated as a handover commitment, even it's not the session boundry on relay chain.